packages/lisk-p2p/src/p2p.ts

```diff
@@ -12,59 +12,191 @@
  * Removal or modification of this copyright notice is prohibited.
  *
  */
-/* tslint:disable:no-unused-expression */
+
+import { EventEmitter } from 'events';
+import http, { Server } from 'http';
+import { platform } from 'os';
+import querystring from 'querystring';
+import socketClusterServer from 'socketcluster-server';
+
+import { Peer } from './peer';
+
 import {
 	NetworkStatus,
-	P2PConfig,
 	P2PMessagePacket,
-	P2PNodeStatus,
-	P2PPenality,
+	P2PPenalty,
 	P2PRequestPacket,
 	P2PResponsePacket,
+	P2PConfig,
+	PeerInfo,
+	RPCResponsePeerList,
+	RPCResponsePeerInfo,
+	P2PNodeStatus,
 } from './p2p_types';
 
 import { PeerPool } from './peer_pool';
 
-export class P2P {
-	private readonly peerPool: PeerPool;
+export const EVENT_NEW_INBOUND_PEER = 'newInboundPeer';
+export const EVENT_FAILED_TO_ADD_INBOUND_PEER = 'failedToAddInboundPeer';
+export const EVENT_NEW_PEER = 'newPeer';
+
+const BASE_10_RADIX = 10;
+
+export class P2P extends EventEmitter {
+	private readonly _config: P2PConfig;
+	private readonly _peerPool: PeerPool;
+	private readonly _httpServer: Server;
+	private readonly _scServer: any;
+	private readonly _newPeers: Set<string>;
+	// TODO: private readonly _triedPeers: Set<string>;
+	private _nodeStatus: P2PNodeStatus;
 
 	public constructor(config: P2PConfig) {
-		this.peerPool = new PeerPool({
-			blacklistedPeers: config.blacklistedPeers,
-			connectTimeout: config.connectTimeout,
-			ipAddress: config.ipAddress,
-			seedPeers: config.seedPeers,
-			wsEngine: config.wsEngine,
+		super();
+		this._config = config;
+		this._nodeStatus = {
 			wsPort: config.wsPort,
-		});
+			os: platform(),
+			version: config.version,
+		};
+		this._newPeers = new Set();
+		// TODO: this._triedPeers = new Set();
+
+		this._peerPool = new PeerPool();
+		this._httpServer = http.createServer();
+		this._scServer = socketClusterServer.attach(this._httpServer);
 	}
 
-	public applyPenalty = (penalty: P2PPenality): void => {
+	public applyPenalty = (penalty: P2PPenalty): void => {
 		penalty;
 	};
 	// TODO
 	public getNetworkStatus = (): NetworkStatus => true;
 	// TODO
-	public getNodeStatus = (): P2PNodeStatus => true;
-	// TODO
-	public request = async (
-		packet: P2PRequestPacket,
-	): Promise<P2PResponsePacket> => {
-		const response = packet;
 
-		return Promise.resolve(response);
-	};
+	public async request<T>(
+		packet: P2PRequestPacket<T>,
+	): Promise<P2PResponsePacket> {
+		return Promise.resolve({ data: packet });
+	}
 
-	public send = (message: P2PMessagePacket): void => {
+	public send<T>(message: P2PMessagePacket<T>): void {
 		message;
 		// TODO
+	}
+
+	public getNodeStatus(): P2PNodeStatus {
```

Comment:
Isn't it better to use getter and setter functions?

Reply:
I also had this thought I but wasn't sure because it deviates from the P2P library API design document https://docs.google.com/document/d/10_LmHqgf7EkPX-U3QIMyyWGBcHcEZGYp82cyfqodQNs/edit# but I guess it is inconsistent with our internal use of getters and setters so I'll make the chance and we can update the document at some point.